src/adapter.rs

```diff
@@ -0,0 +1,525 @@
+
+
+#[cfg(feature = "bluetooth-test")]
+const NOT_SUPPORTED_ON_REAL_ERROR: &'static str = "Error! Test functions are not supported on real devices!";
+#[cfg(feature = "bluetooth-test")]
+const NOT_SUPPORTED_ON_MOCK_ERROR: &'static str = "Error! The first parameter must be a mock structure!";
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::error::Error;
+use bluetooth::BluetoothDevice;
+use bluetooth::BluetoothDiscoverySession;
+
+pub trait BluetoothAdapter {
+    fn get_id(&self)-> String;
+    fn get_devices(&self)-> Result<Vec<BluetoothDevice>, Box<Error>>;
+    fn get_device(&self, address: String) -> Result<Option<BluetoothDevice>, Box<Error>>;
+    fn get_address(&self) -> Result<String, Box<Error>>;
+    fn get_name(&self)-> Result<String, Box<Error>>;
+    fn get_alias(&self) -> Result<String, Box<Error>>;
+    fn get_class(&self)-> Result<u32, Box<Error>>;
+    fn is_powered(&self)-> Result<bool, Box<Error>>;
+    fn is_discoverable(&self) -> Result<bool, Box<Error>>;
+    fn is_pairable(&self)-> Result<bool, Box<Error>>;
+    fn get_pairable_timeout(&self) -> Result<u32, Box<Error>>;
+    fn get_discoverable_timeout(&self)-> Result<u32, Box<Error>>;
+    fn is_discovering(&self)-> Result<bool, Box<Error>>;
+    fn create_discovery_session(&self)-> Result<BluetoothDiscoverySession, Box<Error>> ;
+    fn get_uuids(&self)-> Result<Vec<String>, Box<Error>>;
+    fn get_vendor_id_source(&self)-> Result<String, Box<Error>>;
+    fn get_vendor_id(&self)-> Result<u32, Box<Error>>;
+    fn get_product_id(&self) -> Result<u32, Box<Error>> ;
+    fn get_device_id(&self) -> Result<u32, Box<Error>>;
+    fn get_modalias(&self) -> Result<(String, u32, u32, u32), Box<Error>>;  
+}
+
+impl BluetoothAdapter{
+    #[cfg(all(target_os = "linux", feature = "bluetooth"))]
+    pub fn new() -> Result<Box<BluetoothAdapterBluez>, Box<Error>> {
+        let bluez_adapter = try!(BluetoothAdapterBluez::init());
+        Ok(Box::new(bluez_adapter))
+    }
+
+    #[cfg(all(target_os = "android", feature = "bluetooth"))]
+    pub fn new() -> Result<Box<BluetoothAdapterAndroid>, Box<Error>> {
+        let blurdroid_adapter = try!(BluetoothAdapterAndroid::get_adapter());
+        Ok(Box::new(blurdroid_adapter))
+    }
+
+    #[cfg(all(target_os = "macos", feature = "bluetooth"))]
+    pub fn new() -> Result<Box<BluetoothAdapterMac>, Box<Error>> {
+        let mac_adapter = try!(BluetoothAdapterMac::init());
+        Ok(Box::new(mac_adapter))
+    }
+
+    #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"),
+                  all(target_os = "android", feature = "bluetooth"),
+                  all(target_os = "macos", feature = "bluetooth"))))]
+    pub fn new() -> Result<Box<BluetoothAdapterEmpty>, Box<Error>> {
+        let adapter = try!(BluetoothAdapterEmpty::init());
+        Ok(Box::new(adapter))
+    }
+
+    #[cfg(feature = "bluetooth-test")]
+    pub fn new() -> Result<Box<BluetoothAdapter>, Box<Error>> {
```

Comment:
Since we can have both the bluetooth and the bluetooth-test feature enabled simultaneously, let's call this method `new_mock` instead.

Reply:
Thank you for the feedback!
We have not added a structure for Mock(Arc<BluetoothAdapterMock> yet. Bluetooth-test has additional methods such as _set_id, set_address_ and so on. 
We wanted to confirm if those additional methods can be added in _impl BluetoothAdapter{ }_ instead and have the common methods such as _get_id(), get_address(_).. in _impl BluetoothAdapter for Mock{ }._

Reply:
Mmm, that's interesting. I didn't notice that before. Here is what we should do - add those methods to `trait BluetoothAdapter` and add default implementations that return an Err value. Then for the implementation for the mock adapter, override the defaults with the real implementations. Does that make sense?

Reply:
Yes! Thank you for the clarification. I think we got it. We will make the changes and let you know.